mypy/checkexpr.py

```diff
@@ -1119,13 +1116,25 @@ def check_boolean_op(self, e: OpExpr, context: Context) -> Type:
         self.check_usable_type(left_type, context)
         self.check_usable_type(right_type, context)
 
-        # If either of the type maps is None that means that result cannot happen.
-        # If both of the type maps are None we just have no information.
-        if left_map is not None and right_map is None:
+        if right_map is None:
+            # The boolean expression is statically known to be the left value
+            assert left_map is not None  # find_isinstance_check guarantees this
             return left_type
-        elif left_map is None and right_map is not None:
+        if left_map is None:
+            # The boolean expression is statically known to be the right value
+            assert right_map is not None  # find_isinstance_check guarantees this
             return right_type
-        return UnionType.make_simplified_union([left_type, right_type])
+
+        restricted_left_type = truthiness_restriction(left_type)
+
+        if isinstance(restricted_left_type, UninhabitedType):
+            # The left operand can never be the result
+            return right_type
+        elif left_type is restricted_left_type:
```

Comment:
OK, in that case the latest patch should work for you

On Tue, Aug 16, 2016 at 8:15 PM, David Fisher <notifications@github.com>
wrote:

> In mypy/checkexpr.py
> <https://github.com/python/mypy/pull/1989#discussion_r75001569>:
>
> >              return right_type
> > -        return UnionType.make_simplified_union([left_type, right_type])
> > +
> > +        restricted_left_type = truthiness_restriction(left_type)
> > +
> > +        if isinstance(restricted_left_type, UninhabitedType):
> > +            # The left operand can never be the result
> > +            return right_type
> > +        elif left_type is restricted_left_type:
>
> I think it would be better for the right hand expression to be analyzed
> regardless. Right now there are some places where we don't analyze
> unreachable code, but that's a bug rather than a feature.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/python/mypy/pull/1989/files/19d0920d2f3a5561b12e79ad062a442116edbd40#r75001569>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAMQHSbBd4c_qZ8_cdsk2KF9ru0nzhMJks5qggxqgaJpZM4Jd5oe>
> .
>



-- 
Daniel F. Moisset - UK Country Manager
www.machinalis.com
Skype: @dmoisset